File: 600/628 - Maximum Product of Three Numbers.c
#include <limits.h>

static inline int max(int a, int b) {
    return a > b ? a : b;
}

int maximumProduct(int* nums, int numsSize) {
    int i, mi, m = -4000, ret, max1 = -2000, max2 = -3000, min1 = 2000, min2 = 3000;
    
    if (numsSize == 3) {
        ret = 1;
        for (i = 0; i < 3; ++i)
            ret *= nums[i];
        return ret;
    }
    
    for (i = 0; i < numsSize; ++i) {
        if (nums[i] > m) {
            m = nums[i];
            mi = i;
        }
    }
    
    for (i = 0; i < numsSize; ++i) {
        if (i == mi)
            continue;
        
        if (nums[i] < min1) {
            min2 = min1;
            min1 = nums[i];
        } else if (nums[i] < min2)
            min2 = nums[i];
        
        if (nums[i] > max1) {
            max2 = max1;
            max1 = nums[i];
        } else if (nums[i] > max2)
            max2 = nums[i];
    }

    return max(m * max1 * max2, m * min1 * min2);
}

File: csharp/628.maximum-product-of-three-numbers.cs
/*
 * @lc app=leetcode id=628 lang=csharp
 *
 * [628] Maximum Product of Three Numbers
 */

// @lc code=start
public class Solution {
    public int MaximumProduct(int[] nums) {
        int min1 = int.MaxValue, min2 = int.MaxValue, max1 = int.MinValue, max2 = int.MinValue, max3 = int.MinValue;
        foreach (int n in nums) {
            if (n < min1) {
                min2 = min1;
                min1 = n;
            } else if (n < min2) {
                min2 = n;
            }
            if (n > max1) {
                max3 = max2;
                max2 = max1;
                max1 = n;
            } else if (n > max2) {
                max3 = max2;
                max2 = n;
            } else if (n > max3) {
                max3 = n;
            }
        }
        return Math.Max(max2 * max3 * max1, min1 * min2 * max1);
    }
}
// @lc code=end

